src/shared_utilities/DataHandling.jl

```diff
@@ -0,0 +1,246 @@
+"""
+function available_times(data_handler::DataHandler)
+    return data_handler.available_times
+end
+
+"""
+    available_dates(data_handler::DataHandler)
+
+Return the dates of the snapshots in the data.
+"""
+function available_dates(data_handler::DataHandler)
+    return data_handler.available_dates
+end
+
+"""
+    previous_time(data_handler::DataHandler, time::AbstractFloat)
+    previous_time(data_handler::DataHandler, date::Dates.DateTime)
+
+Return the time in seconds of the snapshot before the given `time`.
+If `time` is one of the snapshots, return itself.
+"""
+function previous_time(data_handler::DataHandler, time::AbstractFloat)
+    time in data_handler.available_times && return time
+    index = searchsortedfirst(data_handler.available_times, time) - 1
+    return data_handler.available_times[index]
+end
+
+function previous_time(data_handler::DataHandler, date::Dates.DateTime)
+    if date in data_handler.available_dates
+        index = searchsortedfirst(data_handler.available_dates, date)
+    else
+        index = searchsortedfirst(data_handler.available_dates, date) - 1
+    end
+    return data_handler.available_times[index]
+end
+
+"""
+    next_time(data_handler::DataHandler, time::AbstractFloat)
+    next_time(data_handler::DataHandler, date::Dates.DateTime)
+
+Return the time in seconds of the snapshot after the given `time`.
+If `time` is one of the snapshots, return the next time.
+"""
+function next_time(data_handler::DataHandler, time::AbstractFloat)
+    index = searchsortedfirst(data_handler.available_times, time)
+    time in data_handler.available_times && (index += 1)
+    return data_handler.available_times[index]
+end
+
+function next_time(data_handler::DataHandler, date::Dates.DateTime)
+    if date in data_handler.available_dates
+        index = searchsortedfirst(data_handler.available_dates, date) + 1
+    else
+        index = searchsortedfirst(data_handler.available_dates, date)
+    end
+    return data_handler.available_times[index]
+end
+
+"""
+    snapshot(data_handler::DataHandler, date::Dates.DateTime)
+    snapshot(data_handler::DataHandler, time::AbstractFloat)
+    snapshot(data_handler::DataHandler)
+
+Return the snapshot from `data_handler` associated to the given `time` (if relevant).
+
+The `time` has to be available in the `data_handler`.
+
+`snapshot` potentially modifies the internal state of `data_handler` and it might be a very
+expensive operation.
+"""
+function snapshot(data_handler::DataHandler, date::Dates.DateTime)
+
+    # Dates.DateTime(0) is the cache key for static maps
+    if date != Dates.DateTime(0)
+        date in data_handler.available_dates || error("date not available")
+    end
+
+    regridder_type = nameof(typeof(data_handler.regridder))
+
+    snapshot = get!(data_handler._cached_regridded_fields, date) do
+        # Add here the arguments when we have more regridders
+        regrid_args = Dict(:TempestRegridder => (date,))
```

Comment:
I am confused why the regridding needs the date.

I am a little confused by the notation also. we get the value of `data_handler._cached_regridded_fields` corresponding do the date, but then what is that called in the `do` block/where is it used in the `do` block? is that what `date` then refers to (instead of the actual date)?

Reply:
TempestRegridder regrids everything ahead of time, saves everything into files, and then read the a filename that is constructed with the date. That's why we need the date.

This block does the following: if `date` is a key in `data_handler._cached_regridded_fields`, then return the associated value. If `date` is not yet a key, then execute the do-block and add the return value as key. So, we change `data_handler._cached_regridded_fields` to add a new `date` with the regridded value.